File: projects/06/quadratic.rb
## quadratic.adb recieves a quadratic equation and returns its roots
##
## Input: a, b, c (numbers)
## Precondition:  a, b, c of form ax^2 + bx + c  
## Output: roots of 'ax^2 + bx + c'
##
# Completed by: Josiah Ryan
# Date: 03/17/21
##########################################################

# getRoots() returns roots of given inputs          
# Receive: a, b, c of form ax^2 + bx + c     
# PRE: a != 0 , and b^2 - 4ac >= 0
# Passback: roots of the inputs when inputed into the above equation
######################################################

def getRoots (a, b, c)
  if a != 0 
    arg = b*b - 4*a*c
    if arg >= 0 
      #returns list with both roots
      [(-b + Math.sqrt(arg)) / (2*a), (-b - Math.sqrt(arg)) / (2*a), 1 ]
    else 
      print "*** getRoots(): b^2 - 4ac is negative!\n";
      [0, 0, 0]
    end
  else 
    print "*** getRoots(): a is zero!\n";
    [0, 0, 0]
  end 
end 


def main
  print "Enter a quadratic equation, enter a, b, c in of the form 'ax^2 + bx +c':\n";
  print "a:" ;
  a = gets.to_i
  print "b:"; 
  b = gets.to_i
  print "c:"; 
  c = gets.to_i
  roots = getRoots(a, b, c) # stores results in 'roots'
  if roots[2] == 1 
    puts "\nThe roots are" 
    puts roots[0] 
    puts roots[1]  
  end 
end

main

File: projects/06/clojure/src/quadratic.clj
;;;; quadratic.clj recieves a quadratic equation and returns its roots.
;;;;
;;;; Input: ints: a, b, c (of form ax^2 + bx + c) 
;;;; Output: the roots when inputed into the above equations
;;;;
;;;; Completed by: Josiah Ryan
;;;; Date: 03/15/21
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns quadratic)      ; name the program

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getRoots() gets roots for quadratic equation 
;;; Receive: ints a, b, c 
;;; PRE: a > 0, (b^2 - 4ac) is not negative
;;; Return: a vector with the roots in it
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn getRoots [a b c]
  (if (not= a 0)
    (let [arg (- (* b b) (* 4 a c))]
      (if (>= arg 0)
        (vector (float (/ (+ (- 0 b) (Math/sqrt arg))(* 2 a))) (float (/ (- (- 0 b) (Math/sqrt arg)) (* 2 a))) 1.0) 
        (do (print "\n*** getRoots(): b^2 - 4ac is negative!\n") (vector 0.0 0.0 0.0)) ; if (arg >= 0)  else
      )
    )
      ;else
    (do  (print "\n*** getRoots(): a is zero!\n")  (vector 0.0  0.0 0.0)) ; else
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; main function
;;; Input: ints a, b, c from 'ax^2 + bx + c'
;;; Output: the roots of the quadratic equation
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn -main []
  (print "Enter a quadratic equation of form 'ax^2 + bx + c': ") (flush)
  (print "\nEnter a:") (flush)
  (let [ a (read) ]                      ; read the string
    (print "Enter b:") (flush)
    (let [ b (read) ]
      (print "Enter c:") (flush)
      (let [ c (read)
             result (getRoots a b c)	 ; calculate the roots
             root1 (get result 0)    	  	 ; store root1
	           root2 (get result 1)		 ; store root1
             valid (get result 2)     ; gets the 3rd item of the vector: 1 if there was no error, 0 if there was
	   ]
      ; display the parts
        (if (= valid 1.0)    ; only displays the roots if valid roots were returned (no errors)
          (printf (str "\nThe roots are: \n %s\n " "%s \n")  root1 root2)
        )
      )
    )
  )
)

